### ISSUE TITLE
DARE cannot solve trivial problem (stuck in infinite loop)

### ISSUE THREAD

**Issue by RussTedrake** [RobotLocomotion/drake#8034]
Calling `DiscreteAlgebraicRiccatiEquation` with 
```
A = [ 0 , 1; 0, 0 ]
B = [ 0; 1 ]
Q = eye(2)
R = 1
```
results in the DARE solver getting stuck in an infinite loop in the `reorder_eigen` subroutine.

**Comment by RussTedrake** [comment]
fwiw - i suspect the right solution here is to fail early (since it's uncontrollable).  i'm actually surprised that matlab's dlqr returns something (`K = [0., 0.], S = diag([1., 2.])`)

**Comment by weiqiao** [comment]
The bug is due to the incorrect assumption that the first generalized eigenvalue of (S,T) is always unstable. It will be fixed soon. The solution is P = diag([1,2]).

**Comment by RussTedrake** [comment]
this also falls into the infinite loop when `A=B=Q=R=eye(2)`

**Comment by weiqiao** [comment]
> this also falls into the infinite loop when `A=B=Q=R=eye(2)`

You were not using the fixed version.

**Comment by RussTedrake** [comment]
Where is the fixed version? I’m running on master...

**Comment by weiqiao** [comment]
> Where is the fixed version? I’m running on master...
The fixed version is here
https://github.com/RobotLocomotion/drake/pull/8469

**Comment by miquelramirez** [comment]
Hi guys,

a brief heads up on this issue. The trivial test case proposed by @RussTedrake is still causing an infinite loop on `master`. Looking through the code, I think that the loop starting on line 

https://github.com/RobotLocomotion/drake/blob/2d7b81fcdd80ce1370670d008d6d86e795a9edf3/math/discrete_algebraic_riccati_equation.cc#L315

is not correct. The problem is that whenever `p` == `q` either pointer do not get increased by `block_size` when they should (I think). A possible fix would be to replace the code in lines 

https://github.com/RobotLocomotion/drake/blob/2d7b81fcdd80ce1370670d008d6d86e795a9edf3/math/discrete_algebraic_riccati_equation.cc#L338-L342

by

```C++
    if (p != q) {
      swap_block(S, T, Z, p, q, q_block_size);
    }
    p += q_block_size;
    q += q_block_size;
```

**Comment by hongkai-dai** [comment]
Resolved in #14865
